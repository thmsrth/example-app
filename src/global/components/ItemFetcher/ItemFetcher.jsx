import React from 'react';

import PropTypes from 'prop-types';
import filtersValuesType from '../Filter/types/filtersValues';

import { getItemById } from '../../utils/FormatHelpers';
import { updateUrlQueryParams } from '../../utils/UrlHelpers';
import I18nGlobal from '../../utils/I18nGlobal';

import NotFound from '../../security/components/NotFound';
import DataStateNotifier from '../DataStateNotifier';

class ItemFetcher extends React.Component {
  componentDidMount() {
    const {
      id, dataPageSize, dataPageNumber,
      dataSortParam, dataFilters, data, fetchDataById,
    } = this.props;

    if (!Number.isNaN(id)) {
      const item = getItemById(id, data);

      if (!item) {
        const defaultParams = {
          size: dataPageSize,
          page: dataPageNumber,
          sort: dataSortParam,
          ...dataFilters,
        };
        const queryParams = updateUrlQueryParams({ id }, defaultParams);

        fetchDataById(queryParams);
      }
    }
  }

  render() {
    const {
      id, data, dataLoading, dataError, resetDataError, children,
    } = this.props;
    const item = getItemById(id, data);

    if (item) {
      return React.Children.map(children, child => React.cloneElement(child, {
        item,
        dataLoading,
        dataError,
        resetDataError,
      }));
    }

    return (
      <DataStateNotifier
        dataLoading={dataLoading}
        dataError={dataError}
        resetDataError={resetDataError}
      >
        <NotFound
          message={I18nGlobal.t('en.route.notFound')}
          loading={dataLoading}
        />
      </DataStateNotifier>
    );
  }
}

export default ItemFetcher;

ItemFetcher.propTypes = {
  dataPageSize: PropTypes.number,
  dataPageNumber: PropTypes.number,
  dataSortParam: PropTypes.string,
  dataFilters: filtersValuesType,
  dataLoading: PropTypes.string,
  dataError: PropTypes.string,
  resetDataError: PropTypes.func,
  fetchDataById: PropTypes.func,
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.array,
  id: PropTypes.number,
  children: PropTypes.node,
};

ItemFetcher.defaultProps = {
  dataPageSize: 25,
  dataPageNumber: 0,
  dataSortParam: 'id,ASC',
  dataFilters: {},
  dataLoading: '',
  dataError: '',
  resetDataError: () => {},
  fetchDataById: () => {},
  id: undefined,
  data: [],
  children: null,
};
