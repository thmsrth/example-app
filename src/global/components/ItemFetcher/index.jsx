import React from 'react';
import { connect, Provider } from 'react-redux';

import store from '../../../app/store';

import {
  fetchDataById,
  resetDataError,
} from '../../../app/actions/data';

import {
  getData,
  getDataError,
  getDataLoading,
  getDataPagination,
  getDataPageSize,
  getDataPageNumber,
  getDataSort,
  getDataSortParam,
  getDataFilters,
} from '../../../app/reducers/data';

import ItemFetcher from './ItemFetcher';

const mapStateToProps = state => ({
  data: getData(state),
  dataError: getDataError(state),
  dataLoading: getDataLoading(state),
  dataPagination: getDataPagination(state),
  dataPageSize: getDataPageSize(state),
  dataPageNumber: getDataPageNumber(state),
  dataSort: getDataSort(state),
  dataSortParam: getDataSortParam(state),
  dataFilters: getDataFilters(state),
});

const mapDispatchToProps = dispatch => ({
  fetchDataById: params => dispatch(fetchDataById(params)),
  resetDataError: () => dispatch(resetDataError()),
});

const ConnectedItemFetcherContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ItemFetcher);

const ItemFetcherContainer = props => (
  <Provider store={store}>
    <ConnectedItemFetcherContainer {...props} />
  </Provider>
);

export default ItemFetcherContainer;
