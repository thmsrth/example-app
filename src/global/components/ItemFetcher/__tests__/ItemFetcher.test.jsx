import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import ItemFetcher from '..';

const TestComponent = () => (<div>Test</div>);

const props = {
  data: [
    {
      id: 1,
      name: 'test',
    },
  ],
  id: 1,
};

describe('ItemFetcher', () => {
  it('should render the child component correctly', () => {
    const tree = shallow(<ItemFetcher {...props}><TestComponent /></ItemFetcher>);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
  it('should render the Not Found correctly', () => {
    const tree = shallow(<ItemFetcher {...props} id={2}><TestComponent /></ItemFetcher>);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
