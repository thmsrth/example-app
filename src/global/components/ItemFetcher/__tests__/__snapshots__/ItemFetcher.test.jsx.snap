// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemFetcher should render the Not Found correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ItemFetcher)
    data={
      Array [
        Object {
          "id": 1,
          "name": "test",
        },
      ]
    }
    id={2}
  >
    <TestComponent />
  </Connect(ItemFetcher)>
</Provider>
`;

exports[`ItemFetcher should render the child component correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ItemFetcher)
    data={
      Array [
        Object {
          "id": 1,
          "name": "test",
        },
      ]
    }
    id={1}
  >
    <TestComponent />
  </Connect(ItemFetcher)>
</Provider>
`;
